### PR TITLE
Update connected players display

### PR DESCRIPTION
**Pull request by Saphs**

### CODE REVIEW

File: src/app/services/board-tiles.service.ts [hunk 1]
```diff
@@ -99,35 +103,23 @@ export class BoardTilesService implements ColyseusNotifyable {
...
+        .subscribe((tiles: Tile[]) => {
+          this.tiles = tiles;
+          observer.next([1, this.tiles.length + 1]);
+          console.info(
+            'Tiles are:',
+            this.tiles.map((t) => t.title)
+          );
+          this._updateFields(observer);
+        })
+        .unsubscribe();
```
Comment (DanielLiebler) [inline src/app/services/board-tiles.service.ts:121]: what is this trying to achieve

File: src/app/services/colyseus-client.service.ts [hunk 2]
```diff
@@ -92,74 +107,80 @@ export class ColyseusClientService {
...
+            MessageType[value[0]] +
+            ' => { \n' +
+            Array.from(value[1].keys())
+              .map((id) => '\t' + id + ' => f(...)')
+              .join(',\n') +
+            '\n}'
+          );
+        })
+        .join(',\n\n')
+    );
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/services/colyseus-client.service.ts:184]: fixed it by introducing a fitting replacement function for changecallbacks as the print structure did not correctly fit when switching to the changeCallbacks.

File: src/app/components/game/interface/connected-players/connected-players.component.html [hunk 3]
```diff
@@ -1,21 +1,59 @@
...
+          <div class="background-layer">
+            <div class="index">{{(i + 1) | asRomanNumeral}}</div>
+          </div>
+          <div class="content-layer">
+            <div class="player-icon">
+              <app-player-icon [borderThickness]="'3px'" [enableUpload]="false" [loginName]="p.loginName" [size]="'2rem'"></app-player-icon>
+              <i *ngIf="p.isCurrentHost" title="Diese Person ist der Spielleiter" class="fas fa-crown host-badge"></i>
+              <i *ngIf="!p.isConnected" title="Diese Person hat das Fenster geschlossen und wird in 2 Minuten raus geworfen" class="fas fa-wifi disconnect-badge"></i>
+              <!-- TODO: Retrieve this value from the DB using roles. Roles should be implemented correctly first -->
+              <i *ngIf="(playerIsDev$ | async).get(p.loginName)" title="Diese Person ist ein Entwickler des Spiels" class="fab fa-dev dev-badge"></i>
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/components/game/interface/connected-players/connected-players.component.html]: added

File: src/app/components/game/interface/connected-players/prescript.pipe.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+@Pipe({
+  name: 'prescript',
+})
+export class PrescriptPipe implements PipeTransform {
+  private static readonly DEFAULT = 'Name';
+  private static readonly LEFT = 'Linker Nachbar';
+  private static readonly RIGHT = 'Rechter Nachbar';
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/components/game/interface/connected-players/prescript.pipe.ts]: adjusted

File: src/app/components/game/viewport/helpers/PhysicsCommands.ts [hunk 5]
```diff
@@ -236,11 +229,10 @@ export class PhysicsCommands implements ColyseusNotifyable {
     rotZ?: number,
     rotW?: number
   ): void {
-    if (entity === PhysicsEntity.figure && this.isPlayerCached(physicsId)) {
+    /*if (entity === PhysicsEntity.figure && this.isPlayerCached(physicsId)) {
       // if playerfigure was already cached dont load it
-      onDone();
       return;
-    }
+    }*/
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/components/game/viewport/helpers/PhysicsCommands.ts]: removed

File: src/app/components/game/viewport/viewport.component.ts [hunk 6]
```diff
@@ -75,10 +77,18 @@ export class ViewportComponent implements AfterViewInit {
 
     this.renderer.render(this.sceneTree, this.camera);
     this.userInteractionController.cameraControls.update();
-    this.stats.update();
+    // this.stats.update();
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/components/game/viewport/viewport.component.ts]: added a comment as it is meant to be re added at a later date

File: src/app/services/board-item-control.service.ts [hunk 7]
```diff
@@ -50,6 +52,32 @@ export class BoardItemControlService implements ColyseusNotifyable {
...
+          console.debug('changing hiddenState', p.hasLeft, figureItem.isHidden, this.allFigures);
+          if (figureItem.isHidden) {
+            this.sceneTree.add(figureItem.mesh);
+            figureItem.isHidden = false;
+          } else {
+            this.sceneTree.remove(figureItem.mesh);
+            figureItem.isHidden = true;
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/services/board-item-control.service.ts]: adjusted

File: src/app/components/game/viewport/viewport.component.ts [hunk 8]
```diff
@@ -42,7 +44,7 @@ export class ViewportComponent implements AfterViewInit {
     public itemService: ItemService,
     private bic: BoardItemControlService
   ) {
-    this.stats = Stats();
+    // this.stats = Stats();
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/components/game/viewport/viewport.component.ts]: added a comment as it is meant to be re added at a later date

File: src/app/services/board-tiles.service.ts [hunk 9]
```diff
@@ -99,35 +103,23 @@ export class BoardTilesService implements ColyseusNotifyable {
...
-    this.tiles = this.gameState.getBoardLayoutAsArray();
-    const tileReadable = [];
-    this.tiles.forEach((tile: Tile) => {
-      if (tile !== undefined) {
-        tileReadable[tile.tileId] = tile.title;
-      }
+  initialize(addToScene: (grp: THREE.Group) => void): Observable<Progress> {
+    return new Observable<Progress>((observer: Observer<Progress>) => {
+      const grp: THREE.Group = this.generateField();
+      addToScene(grp);
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/services/board-tiles.service.ts]: adjusted it

File: src/app/services/colyseus-client.service.ts [hunk 10]
```diff
@@ -28,44 +30,57 @@ export interface CreateRoomOpts {
...
-  private client: Client = new Client(this.backendWsTarget);
-  private activeRoom: BehaviorSubject<Room<GameState>>;
-  private messageCallbacks: Map<MessageType, MessageCallback[]> = new Map<MessageType, MessageCallback[]>([]);
-  private onChangeCallbacks: ((changes: DataChange<any>[]) => void)[] = [this.onDataChange.bind(this)];
-
-  availableRooms: BehaviorSubject<RoomAvailable<RoomMetaInfo>[]>;
+  /** Constants and development parameters */
+  private readonly VERBOSE_CALLBACK_LOGGING = false;
+  private readonly BACKEND_WS_TARGET = environment.wsEndpoint;
+  private readonly client: Client = new Client(this.BACKEND_WS_TARGET);
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/services/colyseus-client.service.ts]: changed it

File: src/app/components/game/game.component.ts [hunk 11]
```diff
@@ -1,80 +1,129 @@
...
+      action: GameActionType.refreshData,
+    };
+    currentRoom.send(MessageType.GAME_MESSAGE, msg);
+  }
+
+  private _handleProgress(p: Progress) {
+    if (p[0] === 0 && p[1] === 0) {
+      this._startLoading();
+    } else if (this.isLoading) {
+      this.loadingScreenRef.setProgress((p[0] / p[1]) * 100);
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/components/game/game.component.ts:96]: you are correct, i added a check

File: src/app/components/game/interface/connected-players/connected-players.component.ts [hunk 12]
```diff
@@ -1,53 +1,152 @@
...
+          map((myLogin: string) => {
+            const myIdx: number = pList.findIndex((p) => p.loginName === myLogin);
+            return this._safeNeighbourIndices(myIdx, pList.length);
+          })
+        );
+      })
+    );
+  }
+
+  private _safeNeighbourIndices(idx: number, listLength: number): [number, number] {
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/components/game/interface/connected-players/connected-players.component.ts:92]: implemented your changes

File: src/app/components/game/interface/next-turn-button/next-turn-button.component.ts [hunk 13]
```diff
@@ -17,37 +17,26 @@ export class NextTurnButtonComponent implements ColyseusNotifyable {
 
   constructor(private gameState: GameStateService) {}
```
[the remark this answers is not in the document]
Comment (Saphs) [reply src/app/components/game/interface/next-turn-button/next-turn-button.component.ts]: done

File: src/app/services/board-tiles.service.ts (hunk 1 above)
Comment (Saphs) [reply src/app/services/board-tiles.service.ts]: nothing - it was meant tin make you think badly of dogs

**Review by manuelstolze — APPROVED** [review]
LGTM!

**Review by DanielLiebler — APPROVED** [review]

**Pull request merged**